starknet-foundry/crates/cheatable-starknet/src/rpc.rs

```diff
@@ -0,0 +1,505 @@
+    }
+}
+// Copied over (with modifications) from blockifier/src/execution/entry_point.rs:144
+fn execute_call_entry_point(
+    entry_point: &mut CallEntryPoint,
+    state: &mut dyn State,
+    cheated_state: &CheatedState,
+    resources: &mut ExecutionResources,
+    context: &mut EntryPointExecutionContext,
+) -> EntryPointExecutionResult<CallInfo> {
+    // Validate contract is deployed.
+    let storage_address = entry_point.storage_address;
+    let storage_class_hash = state.get_class_hash_at(entry_point.storage_address)?;
+    if storage_class_hash == ClassHash::default() {
+        return Err(
+            PreExecutionError::UninitializedStorageAddress(entry_point.storage_address).into(),
+        );
+    }
+
+    let class_hash = match entry_point.class_hash {
+        Some(class_hash) => class_hash,
+        None => storage_class_hash, // If not given, take the storage contract class hash.
+    };
+    // Hack to prevent version 0 attack on argent accounts.
+    if context.account_tx_context.version == TransactionVersion(StarkFelt::from(0_u8))
+        && class_hash
+            == ClassHash(
+                StarkFelt::try_from(FAULTY_CLASS_HASH).expect("A class hash must be a felt."),
+            )
+    {
+        return Err(PreExecutionError::FraudAttempt.into());
+    }
+    // Add class hash to the call, that will appear in the output (call info).
+    entry_point.class_hash = Some(class_hash);
+    let contract_class = state.get_compiled_contract_class(&class_hash)?;
+
+    execute_if_cairo1(
+        entry_point.clone(),
+        contract_class,
+        state,
+        cheated_state,
+        resources,
+        context,
+    )
+    .map_err(|error| {
+        match error {
+            // On VM error, pack the stack trace into the propagated error.
+            EntryPointExecutionError::VirtualMachineExecutionError(error) => {
+                context
+                    .error_stack
+                    .push((storage_address, error.try_to_vm_trace()));
+                // TODO(Dori, 1/5/2023): Call error_trace only in the top call; as it is
+                // right now,  each intermediate VM error is wrapped
+                // in a VirtualMachineExecutionErrorWithTrace  error
+                // with the stringified trace of all errors below
+                // it.
+                EntryPointExecutionError::VirtualMachineExecutionErrorWithTrace {
+                    trace: context.error_trace(),
+                    source: error,
+                }
+            }
+            other_error => other_error,
+        }
+    })
+}
+
+fn execute_if_cairo1(
+    call: CallEntryPoint,
+    contract_class: ContractClass,
+    state: &mut dyn State,
+    cheated_state: &CheatedState,
+    resources: &mut ExecutionResources,
+    context: &mut EntryPointExecutionContext,
+) -> EntryPointExecutionResult<CallInfo> {
+    match contract_class {
+        ContractClass::V0(_) => panic!("Cairo 0 classes are not supported"),
```

Comment:
It will make the old classes behavior a bit unexpected since they cannot be cheated right now - should we consider implementing the cheats for v0 as well?

Reply:
We could create a task to triage to think about it